### PR TITLE
SONARJNKNS-246 Move global configuration to dedicated component

### PR DESCRIPTION
**Pull request by dbmeneses**

### CODE REVIEW

File: src/main/java/hudson/plugins/sonar/SonarGlobalConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    synchronized (this) {
+      if (migrated) {
+        return;
+      }
+      // SonarPublisher might be null if Maven plugin is disabled or not installed
+      DescriptorImpl publisher = Jenkins.getInstance().getDescriptorByType(SonarPublisher.DescriptorImpl.class);
+      if (publisher != null && publisher.getInstallations() != null && publisher.getInstallations().length > 0) {
+
+        if (ArrayUtils.isEmpty(this.installations)) {
+          this.installations = publisher.getInstallations();
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/SonarGlobalConfiguration.java]: MINOR Remove this use of "getInstallations"; it is deprecated. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: src/main/java/hudson/plugins/sonar/SonarGlobalConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+          this.buildWrapperEnabled = publisher.isBuildWrapperEnabled();
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/SonarGlobalConfiguration.java]: MINOR Remove this use of "isBuildWrapperEnabled"; it is deprecated. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: src/main/java/hudson/plugins/sonar/SonarGlobalConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    if (migrated) {
+      return;
+    }
+
+    synchronized (this) {
+      if (migrated) {
+        return;
+      }
+      // SonarPublisher might be null if Maven plugin is disabled or not installed
+      DescriptorImpl publisher = Jenkins.getInstance().getDescriptorByType(SonarPublisher.DescriptorImpl.class);
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/SonarGlobalConfiguration.java]: BLOCKER NullPointerException might be thrown as 'getInstance' is nullable here [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2259)

File: src/main/java/hudson/plugins/sonar/SonarGlobalConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+import org.kohsuke.stapler.StaplerRequest;
+
+import java.util.List;
+
+@Extension(ordinal = 100)
+public class SonarGlobalConfiguration extends GlobalConfiguration {
+  @CopyOnWrite
+  private volatile SonarInstallation[] installations = new SonarInstallation[0];
+  private volatile boolean buildWrapperEnabled = false;
+  private transient boolean migrated = false;
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/SonarGlobalConfiguration.java]: MINOR Remove the "transient" modifier from this field. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2065)

File: src/main/java/hudson/plugins/sonar/SonarGlobalConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+      return;
+    }
+
+    synchronized (this) {
+      if (migrated) {
+        return;
+      }
+      // SonarPublisher might be null if Maven plugin is disabled or not installed
+      DescriptorImpl publisher = Jenkins.getInstance().getDescriptorByType(SonarPublisher.DescriptorImpl.class);
+      if (publisher != null && publisher.getInstallations() != null && publisher.getInstallations().length > 0) {
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/SonarGlobalConfiguration.java]: MINOR Remove this use of "getInstallations"; it is deprecated. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
MINOR Remove this use of "getInstallations"; it is deprecated. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: src/main/java/hudson/plugins/sonar/SonarPublisher.java [hunk 6]
```diff
@@ -441,23 +435,38 @@ public String getDisplayName() {
 
     /**
      * @return all configured {@link hudson.plugins.sonar.SonarInstallation}
+     * @deprecated Global configuration has migrated to {@link SonarGlobalConfiguration}
      */
+    @Deprecated
     public SonarInstallation[] getInstallations() {
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/SonarPublisher.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)

File: src/main/java/hudson/plugins/sonar/SonarPublisher.java [hunk 7]
```diff
@@ -441,23 +435,38 @@ public String getDisplayName() {
...
     }
 
-    public void setInstallations(SonarInstallation... installations) {
+    
+    /**
+     * @deprecated Global configuration has migrated to {@link SonarGlobalConfiguration} 
+     */
+    @Deprecated
+    @VisibleForTesting
+    void setInstallations(SonarInstallation[] installations) {
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/SonarPublisher.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)

File: src/main/java/hudson/plugins/sonar/SonarPublisher.java [hunk 8]
```diff
@@ -441,23 +435,38 @@ public String getDisplayName() {
...
-      save();
     }
 
-    public void setBuildWrapperEnabled(boolean enabled) {
+    /**
+     * @deprecated Global configuration has migrated to {@link SonarGlobalConfiguration} 
+     */
+    @Deprecated
+    @VisibleForTesting
+    void setBuildWrapperEnabled(boolean enabled) {
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/SonarPublisher.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)

File: src/main/java/hudson/plugins/sonar/SonarPublisher.java [hunk 9]
```diff
@@ -422,7 +414,10 @@ public boolean usesPrivateRepository() {
     return usePrivateRepository;
   }
 
-  @Extension(ordinal = 1000)
+  /**
+   * Optional because Maven plugin might not be available.
+   */
+  @Extension(optional = true, ordinal = 1000)
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/SonarPublisher.java]: MINOR Reorder annotation arguments to match the order of declaration. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS3340)

File: src/main/java/hudson/plugins/sonar/SonarPublisher.java [hunk 10]
```diff
@@ -441,23 +435,38 @@ public String getDisplayName() {
...
+    @Deprecated
     public SonarInstallation[] getInstallations() {
       return installations;
     }
 
+    /**
+     * @deprecated Global configuration has migrated to {@link SonarGlobalConfiguration} 
+     */
+    @Deprecated
     public boolean isBuildWrapperEnabled() {
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/SonarPublisher.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)

**Comment by SonarTech** [comment]
SonarQube analysis reported 11 issues:
* BLOCKER 1 blocker
* MINOR 6 minor
* INFO 4 info

Watch the comments in this conversation to review them.

**Comment by henryju** [comment]
LGTM

**Comment by SonarTech** [comment]
SonarQube analysis reported 11 issues:
* BLOCKER 1 blocker
* MINOR 6 minor
* INFO 4 info

Watch the comments in this conversation to review them.

**Pull request closed without merging**